### ISSUE TITLE
Adding Batch Size as explicit parameter for Batch Normalization layer

### ISSUE THREAD

**Issue by pranv** [fchollet/keras#61]
I think it would be much more clear and easy to have a "batch size"  parameter separately for Batch Normalization layer.  We can just directly pass the outputs of our convolution or pooling layers to it. The layers as a whole will be more coherent. 
(As an aside,  did anyone have any luck with batch normalization?  I tried many times,  but actually got worse results most of the time.)

**Comment by fchollet** [comment]
As an example, batch normalization seems to have a significant positive impact on the Reuters example. 

Batch size during training is assumed to not be fixed, and may vary when iterating over the data (in particular, if you use the fit() method of models.Sequential, the last batch will generally be smaller). How would a batch_size layer parameter work?

I also don't see the argument for introducing one. Can you expand on that?

**Comment by pranv** [comment]
Good to hear about the Reuters example. 

Regarding my argument :
What you said is exactly what I'm trying to point out. There are 2 separate things here:

1. Batch Normalization layer in practice would look something like "Model.add(BatchNormalization(batch_size, stack_size,  nb_rows,  nb_columns)).  The last three parameters are something we calculate,  and they are similar to the parameters of other layers.  So,  I just suggested if it would be better to have it has BatchNormalization(stack_size,  nb_rows,  nb_columns, batch_size=32). Not much difference between them,  but. I thought, since it is "batch"  normalization,  it would more explicit and clear. 
2. Like you said,  the last batch size < batch_size usually.  But a batch normalization layer,  with batch size passed in,  to the constructor will give you an error if a batch of lesser items is fed into the network .

**Comment by fchollet** [comment]
The current implement is not dependent on any batch size. The input_shape parameter is assumed to be the shape of a __single__ sample, not of a batch. It seems that this where the confusion originates.

This parameter input_shape is used to determine the shape of the internal gamma and beta parameters, which are used as such:
```
out = self.gamma * X_normed + self.beta
```
Here X_normed has shape (batch_size,) + input_shape (which is the shape of the batch fed as input), and gamma and beta have shape input_shape. Theano handles these operations like numpy would.

I will update the documentation to reflect this.

**Comment by pranv** [comment]
You're right
